rfd/0159-webasset-cache.md

```diff
@@ -135,30 +103,35 @@ if strings.HasPrefix(r.URL.Path, "/web/app") {
 		http.StripPrefix("/web", fs).ServeHTTP(w, r)
         return
 	}
-    res, err := h.auth.proxyClient.GetWebasset(cfg.Context, &webassetcache.GetWebassetRequest{
-		Name: r.URL.Path[len("/web/app/"):],
-	})
+	writer := &MemoryWriterAt{}
+	cfg.WebassetHandler.Download(cfg.Context, session.ID(fmt.Sprintf("%s/%s", teleport.Version, fileName)), writer)
+
 	// detect content type
 	// compress
-	// store the file in this proxy's webasset cache to prevent future requests
-	http.ServeContent(w, r, "example.txt", time.Now(), compressedFileHere)
+	http.ServeContent(w, r, fileNameToSkip, time.Now(), bytes.NewReader(writer.buffer.Bytes()))
 }
 ```
 
 
-### Security Concerns
-We aren't serving any files directly from a file system and only fetching from 
-a map so there shouldn't be any worry about retrieving files that aren't in the map.
-Additionally, only the internal proxy role should be able to set/get these files
-in the map and if a proxy has been compromised, this webasset cache is the least
-of our concerns.
-
+### Security
 This feature should be **opt in only** via configuration. 
 
-Another thing to consider is not allowing this cache to blow up and run OOM. I
-don't currently see any non-arbitrary solution to this. For example, we could
-store the webassets by _version_ rather than just file and limit it, for
-example, to two versions. But then what happens if a customer wants to run 3
-versions? How many versions do we allow? I think making this an explicit choice
-by the customer to turn on should prevent any sort of "accidental" misuse enough
-for now.
\ No newline at end of file
+As of now, only the Auth talks to s3/aws. This means that the proxy would also have to have access to some credentials
+that write/read access to the user's chosen storage. This increases the risk surface area by having two susceptible services. 
+
+If someone has gained access to the bucket, they could potentially overwrite files with their own code that the
```

Comment:
You could take a similar approach to how Teleport handles session recordings.

We enable versioning on the bucket, and when querying we only ever look at the first version.

This means someone who compromises S3 credentials can write a new version of an asset, but Teleport will never serve it because it's only ever going to serve the original version.

Reply:
Interesting. I'll check this out, thanks